android/java/org/chromium/chrome/browser/crypto_wallet/util/TokenUtils.java

```diff
@@ -12,40 +12,88 @@
         ArrayList<ErcToken> arrayTokens = new ArrayList<>(Arrays.asList(tokens));
-        Utils.removeIf(arrayTokens, t -> TokenUtils.shouldTokenBeFilteredOut(t));
+        if (allTokens) {
+            Utils.removeIf(arrayTokens, t -> TokenUtils.shouldTokenBeFilteredOut(t));
+        } else {
+            Utils.removeIf(arrayTokens, t -> TokenUtils.shouldUserVisibleTokenBeFilteredOut(t));
+        }
 
         return arrayTokens.toArray(new ErcToken[0]);
     }
 
     public static void getUserAssetsFiltered(BraveWalletService braveWalletService, String chainId,
             BraveWalletService.GetUserAssetsResponse callback) {
         braveWalletService.getUserAssets(chainId, (ErcToken[] tokens) -> {
-            ErcToken[] filteredTokens = filterOut(tokens);
+            ErcToken[] filteredTokens = filterOut(tokens, false);
             callback.call(filteredTokens);
         });
     }
 
-    public static void getAllTokensFiltered(
-            ErcTokenRegistry ercTokenRegistry, ErcTokenRegistry.GetAllTokensResponse callback) {
+    public static void getAllTokensFiltered(BraveWalletService braveWalletService,
+            ErcTokenRegistry ercTokenRegistry, String chainId,
+            ErcTokenRegistry.GetAllTokensResponse callback) {
         ercTokenRegistry.getAllTokens((ErcToken[] tokens) -> {
-            ErcToken[] filteredTokens = filterOut(tokens);
-            callback.call(filteredTokens);
+            braveWalletService.getUserAssets(chainId, (ErcToken[] userTokens) -> {
+                ErcToken[] filteredTokens =
+                        filterOut(concatenateTwoArrays(tokens, userTokens), true);
+                callback.call(filteredTokens);
+            });
         });
     }
 
     public static void getBuyTokensFiltered(
             ErcTokenRegistry ercTokenRegistry, ErcTokenRegistry.GetAllTokensResponse callback) {
         ercTokenRegistry.getBuyTokens((ErcToken[] tokens) -> {
-            ErcToken[] filteredTokens = filterOut(tokens);
+            ErcToken[] filteredTokens = filterOut(tokens, true);
             callback.call(filteredTokens);
         });
     }
+
+    public static void isCustomToken(ErcToken token, ErcTokenRegistry ercTokenRegistry,
+            org.chromium.mojo.bindings.Callbacks.Callback1<Boolean> callback) {
+        ercTokenRegistry.getAllTokens((ErcToken[] tokens) -> {
+            boolean isCustom = true;
+            for (ErcToken tokenFromAll : tokens) {
+                if (token.contractAddress.equals(tokenFromAll.contractAddress)) {
+                    isCustom = false;
+                    break;
+                }
+            }
+            callback.call(isCustom);
+        });
+    }
+
+    private static ErcToken[] concatenateTwoArrays(ErcToken[] arrayFirst, ErcToken[] arraySecond) {
+        List<ErcToken> both = new ArrayList<>();
+
+        Collections.addAll(both, arrayFirst);
+        for (ErcToken tokenSecond : arraySecond) {
```

Comment:
if we may have a lot of elements at `arraySecond`, maybe use `TreeSet` with custom comparator?

Reply:
I don't think there are will be much custom tokens